Implementation Plan: Additional search views

obulat opened the pull request:
<!-- prettier-ignore -->
## Fixes
<!-- If PR doesn't fully resolve the issue, replace 'Fixes' below with 'Related to'. -->
<!-- If there is no issue being resolved, please consider opening one before creating this pull request. -->

Resolves #2560 by @zackkrida

## Assigned reviewers
- [x] @zackkrida
- [ ] @sarayourfriend

## Note
A very rough draft implementation of the additional search view pages is in branch [draft/collection_pages](https://github.com/WordPress/openverse/tree/draft/collection_pages).

## Current round

> **Note**
> This discussion is following the Openverse decision-making process. Information about this process can be found
on the [Openverse documentation site](https://docs.openverse.org/meta/decision_making/index.html). Requested reviewers or participants will be following this process. If you are being asked to give input on a specific detail, you do not need to familiarise yourself with the process and follow it.

The discussion is currently in the [Decision round](https://docs.openverse.org/meta/decision_making/process_description.html#decision-round).

The deadline for review of this round is **2023-07-25**.

zackkrida commented on documentation/projects/proposals/additional_search_views/20230719-implementation_plan_additional_search_views.md:
```diff
@@ -0,0 +1,263 @@
+
+- items with the selected tag
+- items from the selected source (provider)
+- items by the selected creator. The pages re-use the existing image grid and
```
```suggestion
- items by the selected creator. 

These pages re-use the existing image grid and
```

zackkrida commented on documentation/projects/proposals/additional_search_views/20230719-implementation_plan_additional_search_views.md:
```diff
@@ -0,0 +1,263 @@
+- items by the selected creator. The pages re-use the existing image grid and
+  audio collection views and can load more images or audio items on the Load
+  more button click. The single result pages are updated to add the links to
+  these pages, as seen in the Figma mockups: [link to Figma mockup]
```
```suggestion
  these pages, as seen in the Figma mockups:
  
- [Image single result](https://www.figma.com/file/niWnCgB7K0Y4e4mgxMrnRC/Additional-search-views?type=design&node-id=1200-63284&mode=dev)
- [Audio single result](https://www.figma.com/file/niWnCgB7K0Y4e4mgxMrnRC/Additional-search-views?type=design&node-id=1200-63285&mode=dev)
```

zackkrida commented on documentation/projects/proposals/additional_search_views/20230719-implementation_plan_additional_search_views.md:
```diff
@@ -0,0 +1,263 @@
+
+### Page changes
+
+#### Create a page for `tag` / `creator` /`source` collections.
```
This section is excellent! This should be a great approach

zackkrida commented on documentation/projects/proposals/additional_search_views/20230719-implementation_plan_additional_search_views.md:
```diff
@@ -0,0 +1,263 @@
+search result page because it is a part of the `audio.vue` page. We should
+extract the part that shows the loading skeleton, the column of `VAudioTrack`
+rows and the Load more section. This component will be reused in the audio
+search page and on the Additional search views. One thing that I'm not sure
+about here is whether the analytics events should stay the same and be sent from
+all pages - or if we should somehow differentiate the events coming from
+different search views.
```
> One thing that I'm not sure
about here is whether the analytics events should stay the same and be sent from
all pages - or if we should somehow differentiate the events coming from
different search views.

These will be naturally differentiated by the built-in `pathname` property of all analytics events. They can remain the same.

zackkrida commented on documentation/projects/proposals/additional_search_views/20230719-implementation_plan_additional_search_views.md:
```diff
@@ -0,0 +1,263 @@
+#### Add a `VCollectionHeader` component
+
+Figma links:
+[creator desktop](https://www.figma.com/file/niWnCgB7K0Y4e4mgxMrnRC/Additional-search-views?type=design&node-id=1200-56323&mode=design&t=pN0PPAzlKQEKT9tJ-4),
+[creator mobile](https://www.figma.com/file/niWnCgB7K0Y4e4mgxMrnRC/Additional-search-views?type=design&node-id=1200-56362&mode=design&t=suLIyJHNmZrM0mPH-4),
+[source desktop](https://www.figma.com/file/niWnCgB7K0Y4e4mgxMrnRC/Additional-search-views?type=design&node-id=1200-56381&mode=design&t=suLIyJHNmZrM0mPH-4),
+[source mobile](https://www.figma.com/file/niWnCgB7K0Y4e4mgxMrnRC/Additional-search-views?type=design&node-id=1200-56421&mode=design&t=suLIyJHNmZrM0mPH-4),
+[tag desktop](https://www.figma.com/file/niWnCgB7K0Y4e4mgxMrnRC/Additional-search-views?type=design&node-id=1216-58402&mode=design&t=suLIyJHNmZrM0mPH-4),
+[tag mobile](https://www.figma.com/file/niWnCgB7K0Y4e4mgxMrnRC/Additional-search-views?type=design&node-id=1200-56457&mode=design&t=suLIyJHNmZrM0mPH-4)
```
```suggestion

- [creator desktop](https://www.figma.com/file/niWnCgB7K0Y4e4mgxMrnRC/Additional-search-views?type=design&node-id=1200-56323&mode=design&t=pN0PPAzlKQEKT9tJ-4),
- [creator mobile](https://www.figma.com/file/niWnCgB7K0Y4e4mgxMrnRC/Additional-search-views?type=design&node-id=1200-56362&mode=design&t=suLIyJHNmZrM0mPH-4),
- [source desktop](https://www.figma.com/file/niWnCgB7K0Y4e4mgxMrnRC/Additional-search-views?type=design&node-id=1200-56381&mode=design&t=suLIyJHNmZrM0mPH-4),
- [source mobile](https://www.figma.com/file/niWnCgB7K0Y4e4mgxMrnRC/Additional-search-views?type=design&node-id=1200-56421&mode=design&t=suLIyJHNmZrM0mPH-4),
- [tag desktop](https://www.figma.com/file/niWnCgB7K0Y4e4mgxMrnRC/Additional-search-views?type=design&node-id=1216-58402&mode=design&t=suLIyJHNmZrM0mPH-4),
- [tag mobile](https://www.figma.com/file/niWnCgB7K0Y4e4mgxMrnRC/Additional-search-views?type=design&node-id=1200-56457&mode=design&t=suLIyJHNmZrM0mPH-4)
```

zackkrida commented on documentation/projects/proposals/additional_search_views/20230719-implementation_plan_additional_search_views.md:
```diff
@@ -0,0 +1,263 @@
+
+#### Update the "by" area of the single result page
+
+The "by creator" line should be replaced with a section that is
+horizontally-scrollable on mobile, and contains the `VCreatorLink` and
+`VSourceLink` components.
```
I'd recommend we implement scroll-snap here, as a nice simple enhancement for narrow desktop viewports but also the mobile experience. Example here: 

https://play.tailwindcss.com/AbfA33Za50

zackkrida commented on documentation/projects/proposals/additional_search_views/20230719-implementation_plan_additional_search_views.md:
```diff
@@ -0,0 +1,263 @@
+
+#### Update links in the "information" section
+
+Add external link icon to the provider link; add creator link with external link
+icon The links to creator in the image and audio single result pages should have
+an "external link" icon.
+[image creator link](https://github.com/WordPress/openverse/blob/35f08e5710d59f6c10f0cf54103fcce151adfefe/frontend/src/pages/image/_id/index.vue#L62-L73)
+and
+[audio creator link](https://github.com/WordPress/openverse/blob/35f08e5710d59f6c10f0cf54103fcce151adfefe/frontend/src/components/VAudioTrack/layouts/VFullLayout.vue#L32-L41).
+Audio creator link should also be updated to match the image creator link. It
+should be a conditional component: `VLink` if the `creator_url` is truthy and
+`span` if the `creator_url` is falsy.
+
+Currently, the `foreign_landing_url` is linked to the "source" in the image page
+and "provider" in the audio page. The audio page should be updated to match the
+image page: the `foreign_landing_url` link should be added to the "source", not
+provider.
+[Audio provider link](https://github.com/WordPress/openverse/blob/35f08e5710d59f6c10f0cf54103fcce151adfefe/frontend/src/components/VAudioDetails/VAudioDetails.vue#L61C1-L63),
+[image source link](https://github.com/WordPress/openverse/blob/35f08e5710d59f6c10f0cf54103fcce151adfefe/frontend/src/components/VImageDetails/VImageDetails.vue#L26-L28)
```
This section just needs a bit of formatting, and cleanup it's a little hectic.

zackkrida commented on documentation/projects/proposals/additional_search_views/20230719-implementation_plan_additional_search_views.md:
```diff
@@ -0,0 +1,263 @@
+
+### Analytics changes
+
+I am not sure what events we want to add, if any. Should we add `<VIEW>_CLICKED`
+events for the new views, or would page views be sufficient?
+
+Should we add a new `SELECT_COLLECTION_ITEM` event similar to
+`SELECT_SEARCH_RESULT`?
```
I personally think page views will be sufficient. The only way to access the pages is directly or via links on the single results, which will all be captured by standard page visits.

sarayourfriend replied:
Agreed. `SELECT_SEARCH_RESULT` can be narrowed to `/search` or `/tag*` (for example) to determine where the event occurred'

zackkrida commented on documentation/projects/proposals/additional_search_views/20230719-implementation_plan_additional_search_views.md:
```diff
@@ -0,0 +1,263 @@
+
+The clicks on external creator link in the `VCollectionHeader` should be tracked
+as `VISIT_CREATOR_LINK` events. We don't have a special event for visiting
+source, should we add one, or should they be tracked as generic
+`VISIT_EXTERNAL_LINK`?
```
It makes sense to me to add an event for these because we might want to know how these links are used broadly ("how many provider link clicks are there?") rather than specifically ("how many clicks to https://flickr.com are there?").

zackkrida commented on documentation/projects/proposals/additional_search_views/20230719-implementation_plan_additional_search_views.md:
```diff
@@ -0,0 +1,263 @@
+
+We should add visual-regression tests for the new views. To minimize flakiness
+due to slow loading of the images, we should probably use the
+`{ filter: brightness(0%); }` trick for the images on the page.
```
Are we using this anywhere presently?

obulat replied:
Link added.

zackkrida commented on documentation/projects/proposals/additional_search_views/20230719-implementation_plan_additional_search_views.md:
```diff
@@ -0,0 +1,263 @@
+
+No new tools or packages are necessary.
+
+### Other projects or work
+
+<!-- Note any projects this plan is dependent on. -->
```
```suggestion
### Other projects or work

<!-- Note any projects this plan is dependent on. -->
Not applicable.
```

zackkrida commented on documentation/projects/proposals/additional_search_views/20230719-implementation_plan_additional_search_views.md:
```diff
@@ -0,0 +1,263 @@
+The work on components (`VCreatorLink`, `VSourceLink`, `VMediaTag`), and on the
+search store changes can be parallelized.
+
+We should create a separate branch for this work because adding the links to the
+collection pages on the single result pages (`VCreatorLink`, `VSourceLink`,
+`VMediaTag`) before the collection pages are ready is not possible.
```
Should it be feature-flagged instead, Olga? I would think so. If you agree please update the plan to reflect that.

sarayourfriend replied:
Strongly agree on feature-flagging rather than feature-branching whenever possible.

zackkrida commented on documentation/projects/proposals/additional_search_views/20230719-implementation_plan_additional_search_views.md:
```diff
@@ -0,0 +1,263 @@
+
+<!-- How do we roll back this solution in the event of failure? Are there any steps that can not easily be rolled back? -->
+
+If we work on a separate branch, it would be easy to rollback by reverting the
+branch merge commit.
```
Once again I'd suggest a feature flag rather than a branch-based approach.

zackkrida reviewed (changes requested):
This is excellent, I'm only requesting changes based on the volume of cleanup suggestions I've made.

Actually, one potential blocker. I believe we need to include the necessary API changes in this plan, rather than have a separate plan for the API. Since the API changes are so small.

sarayourfriend commented on documentation/projects/proposals/additional_search_views/20230719-implementation_plan_additional_search_views.md:
```diff
@@ -0,0 +1,263 @@
+
+#### Create a page for `tag` / `creator` /`source` collections.
+
+Nuxt allows creating nested dynamic routes like
+`/pages/_collection/_mediaType/_term`.
+
+Here, the collection can be `tag`, `creator` or `source`. `mediaType` can be one
+of the supported media types (currently, `image` and `audio`). `term` refers to
+the actual value of the tag, creator or source name.
```
That's fantastic! It could be useful for SEO as well. Speaking of which, what considerations do we need to apply to `robots.txt` for this change, if any? I don't really understand our current approach to robots.txt, is it leftover from when we were iframed? @zackkrida I believe you may know.

zackkrida replied:
Our current robots.txt is blank, which allows all pages of the site to be crawled and indexed. That works exactly the same as:

```
User-agent: *
Disallow:
```


The staging site is blocked from indexing. 

I don't think we need to do anything here. These pages can be crawled and indexed.

sarayourfriend replied:
These new pages are essentially search views with static URLs though (not query params). Should that kind of thing be indexed?

Sorry if this is already discussed somewhere. Might be helpful to document the rationale of our robots.txt in the docs site.

sarayourfriend commented on documentation/projects/proposals/additional_search_views/20230719-implementation_plan_additional_search_views.md:
````diff
@@ -0,0 +1,263 @@
+make sure that the `collection`, `searchType` and `term` `params` are valid.
+
+```
+`validate({ params, $pinia }): {
````
Just removing a stray `` ` ``:

```suggestion
validate({ params, $pinia }): {
```

Is this code sample missing a portion of the intended sample?

sarayourfriend commented on documentation/projects/proposals/additional_search_views/20230719-implementation_plan_additional_search_views.md:
```diff
@@ -0,0 +1,263 @@
+tag/creator/source. For source and the creator, there should be an external link
+button.
+
+The header should also display the number of results, "251 audio files with the
+selected tag", "604 images provided by this source", "37 images by this creator
+in Hirshhorn Museum and Sculpture Garden".
```
Is it at all relevant for us to consider sources that have a single creator? Boston Public Library, as a source, for example, only has Boston Public Library as the "creator" of the work. It's weird to think that there would be two separate pages for them, but then I'm not sure how we would communicate this to the user if we automatically redirected the BPL creator search to the BPL source search :thinking:

obulat replied:
Added a note to the plan. By the way, we don't have a Boston Public Library as a source filter on the frontend. What is the link to their API or collection?

sarayourfriend replied:
Oops, I've just realised that BPL is just a creator we've picked up through Flickr, not an actual source. Maybe an easy source to add then!

sarayourfriend commented on documentation/projects/proposals/additional_search_views/20230719-implementation_plan_additional_search_views.md:
```diff
@@ -0,0 +1,263 @@
+
+### Store changes
+
+#### Update the `searchBy` filter in the `search` store
+
+Currently, this filter is shaped as other filters: it is a list of objects with
+`code`, `name` and `checked` properties. This is a possible value now: `{
+searchBy: { code: "creator", name: "filters.searchBy.creator", checked: true }}.
+
+Since this filter is mutually exclusive (you cannot search by both the creator
+and the tag, for example), this shape is very difficult to update.
+
+It is better to set the `searchBy` filter to have one of the possible values:
+`{ searchBy: <null|creator|source|tag> }`.
+
+We should also create a new method, `setSearchBy`, in the `search` store, that
+would allow directly set the `searchBy` filter value.
```
Can you clarify whether this is just a DevEx/code-quality refactor or if it's necessary for the new features? I don't fully understand why this particular change is relevant to the creator/tag/source pages. Are they going to use the same search store and filters? If so, that would be helpful background to have somewhere as well (may have missed it though?).

obulat replied:
Yes, I actually re-used the same search store in the draft implementation that I linked in the issue description. But to get the store to handle different `searchBy` values, we should add this change.
Thank you for asking this question! I realized that I didn't write anything about the data side: what API calls will be made, how the stores will be used and how they need to be updated. I'll add this during the revision round.

sarayourfriend commented on documentation/projects/proposals/additional_search_views/20230719-implementation_plan_additional_search_views.md:
```diff
@@ -0,0 +1,263 @@
+
+<!-- What risks are we taking with this solution? Are there risks that once taken can’t be undone?-->
+
+The biggest risk I see is that this project might be seen as an "invitation" to
+scraping. Hopefully, the work on providing the dataset would minimize such
+risks.
```
Frontend rate limiting will also help.

obulat replied:
Added the note on rate limiting.

sarayourfriend reviewed (commented):
I agree with Zack, but will suggest that either the API changes should be included in this IP or they should be completed first.

Without having a deep knowledge of how the frontend stores work, and because it isn't described in this implementation plan, it's very hard to understand how the searches for these collection pages is meant to be implemented.

sarayourfriend commented on documentation/projects/proposals/additional_search_views/20230719-implementation_plan_additional_search_views.md:
```diff
@@ -0,0 +1,366 @@
+
+<!-- Describe the implementation step necessary for completion. -->
+
+### API changes
```
Thank you for adding this section. It helps a lot to understand how this will all fit together.

I like the approach to add new routes on the API rather than trying to stuff everything into the existing search endpoint. That makes things a lot nicer for the creator and provider especially.

However, there are some problems with using the database rather than ES to access anything:

1. The database does not cache queries in the same way that ES does. Repeated queries will not necessarily be as efficient as from ES.
2. The database does not score documents at all, so the order will different dramatically to the way that ES would order the documents. That's an issue with respect to popularity data today already, but will become even more of an issue if we start to score documents based on other metrics as theorised by our search relevancy discussions.
3. `creator` is not indexed in the API database, so a query against it will be very slow.

Only one of these can be "fixed", the third, but it's much easier to add a `creator.raw` field to the index instead, like we've done for title, description, and tag name. Adding a new field to the index is far easier than adding a new index to the database (which would require careful application to avoid a long-running migration) and which could significantly increase the time it takes to do a data refresh.

I'd also strongly recommend against the route described as it is harder to cache in a sensible way (on the Cloudflare side of things) or to perform cache invalidation that might be required by #1969. Instead of using query strings, we can describe the resource via the path: `/<media type>/source/<source>/creator/<creator>` is very clean, easy to read and understand, and very easy to manage the cache for because it is a static path. The source page can use the same route by leaving off the creator. This removes the need to manage specific query params as well and would allow us to add querying within these routes more easily in the future behind the regular `q` parameter if we wanted.

For the tag route, I'd recommend the singular `tag` rather than plural `tags`, for legibility if we are presenting a single tag. Again, using a route-based approach instead of query parameter approach makes it easier to manage the cache if needed and could allow us to more easily feature specific tags with high quality results as example pages if we wanted, like `/images/tag/dog` or `/images/tag/flying%20birds`.

Path elements still need to be URL encoded to preserve special characters and spaces, but for most cases like `/images/source/nasa`, `/images/sources/nypl`, `/images/sources/inaturalist`, etc we've suddenly got a very good-looking URL to point people to.

sarayourfriend commented on documentation/projects/proposals/additional_search_views/20230719-implementation_plan_additional_search_views.md:
````diff
@@ -0,0 +1,366 @@
+`searchBy` filter value to construct the API query:
+
+```typescript
+const prepareSearchQuery = (
+  searchParams: Record<string, string>,
+  mediaType: SupportedMediaType
+) => {
+  if (searchParams.searchBy) {
+    if (searchBy === "creator") {
+      // the name of the source is stored in the `<mediaType>Provider` filter
+      return {
+        creator: searchTerm,
+        source: searchParams[`${mediaType}Provider`],
+      }
+    } else if (searchBy === "source") {
+      return { source: searchParams[`${mediaType}Provider`] }
+    } else if (searchBy === "tag") {
+      // The parameter is plural in the API
+      return { tags: searchTerm }
+    }
+  } else {
+    // Current search query transform
+  }
+}
````
Just noting that regardless of whether we use my fully route based approach above, we'll still need to determine the path pattern to use. Should this function also handle that, maybe returning either an array of path segments (including holes for parameters and a final object for query parameters if still needed) or the query parameter object to send to the default `search` route?

sarayourfriend commented on documentation/projects/proposals/additional_search_views/20230719-implementation_plan_additional_search_views.md:
```diff
@@ -0,0 +1,366 @@
+
+#### Create a page for `tag` / `creator` /`source` collections.
+
+Nuxt allows creating nested dynamic routes like
+`/pages/_collection/_mediaType/_term`.
+
+Here, the collection can be `tag`, `creator` or `source`. `mediaType` can be one
+of the supported media types (currently, `image` and `audio`). `term` refers to
+the actual value of the tag, creator or source name.
```
Realising now that for these to work, creator needs to have source included. It could make sense to make these match the API routes: `/_mediaType/source/_source/creator/_creator`, `/_mediaType/tag/_tag` etc.

sarayourfriend commented on documentation/projects/proposals/additional_search_views/20230719-implementation_plan_additional_search_views.md:
```diff
@@ -0,0 +1,366 @@
+
+[Figma designs in the dev mode](https://www.figma.com/file/niWnCgB7K0Y4e4mgxMrnRC/Additional-search-views?node-id=22%3A13656&mode=dev)
+
+## Parallelizable streams
```
Should this be updated to include the API work in some way?

sarayourfriend reviewed (changes requested):
I have only three truly blocking issues, the rest are just additional questions or suggestions. The blockers are:

1. The various issues I've described with using Postgres to query for the creator view must be addressed. I've suggested we continue to use ES by adding new `.raw` sub-fields for creator and source.
2. The frontend routes need to disambiguate creators by source. I've suggested alternative paths that mirror my suggested path (rather than query param) based approach for the API and would solve this problem.
3. For me to fully understand _how_ the plan would be implemented, I need an ordered, high-level step-by-step implementation of all the tasks that would be converted into individual issues. I think we should clarify this in the implementation plan template to have something like `ordered steps` and `step descriptions` so that it's possible to get both a high-level view of the plan separated from the detailed descriptions and nuances of each step.

Overall the plan looks very good though. The first two issues are critical blockers for the plan because the plan will not succeed without addressing them. The last is just a personal blocker for me to get a proper understanding of the plan.

I'm starting AFK tomorrow and am confident that you and Zack will be able to resolve the blockers without me. The plan otherwise is great. Please consider this plan approved by me once those blockers are resolved :rocket:

obulat commented:
I've updated the plan to add the API and Elasticsearch changes. In particular, I updated the paths to match @sarayourfriend's suggestions. I also rewrote the plan to make the steps descriptions and steps sequence clearer.
@zackkrida, this is ready for re-review.

zackkrida commented:
I'll review this today!

zackkrida commented on documentation/projects/proposals/additional_search_views/20230719-implementation_plan_additional_search_views.md:
```diff
@@ -0,0 +1,429 @@
+
+<!-- What hard blockers exist which might prevent further work on this project? -->
+
+The main blocker could be the maintainer capacity.
```
This is always a potential blocker, so I don't think we need to explicitly mention it.

zackkrida commented on documentation/projects/proposals/additional_search_views/20230719-implementation_plan_additional_search_views.md:
```diff
@@ -0,0 +1,429 @@
+
+<!-- Choose two people at your discretion who make sense to review this based on their existing expertise. Check in to make sure folks aren't currently reviewing more than one other proposal or RFC. -->
+
+- [ ] @zackkrida
```
```suggestion
- [x] @zackkrida
```

zackkrida commented:
We're going to merge this on the basis of @sarayourfriend's blockers being resolved 🥳
